cmd/snap-recovery-chooser/main.go

```diff
@@ -0,0 +1,198 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// The `snap-recovery-chooser` acts as a proxy between the chooser UI process
+// and the actual snapd daemon.
+//
+// It obtains the list of seed systems and their actions from the snapd API and
+// passed that directly to the standard input of the UI process. The UI process
+// is expected to present the list of options to the user and print out a JSON
+// object with the choice to its standard output.
+//
+// The JSON object carrying the list of systems is the client.ChooserSystems
+// structure. The response is defined as follows:
+// {
+//     "label": "<system-label",
+//     "action": {} // client.SystemAction object
+// }
+//
+// No action is forwarded to snapd if the chooser UI exits with an error code or
+// the response structure is invalid.
+//
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+)
+
+var (
+	// default marker file location
+	defaultMarkerFile = "/run/snapd-recovery-chooser-triggered"
```

Comment:
Mild preference for `/run/snapd/...` for tidiness.

Reply:
This is already baked in a couple of places, including subiquity.